### PR TITLE
fix: path constraints expression (h) is enforced at initial state

### PR DESCRIPTION
**Pull request by mohammed-hababeh**
fix: path constraints expression (h) is enforced at initial state
Bug description:
when using con_h_expr to set path constraints for an OCP, these constraints are enforced at nodes (0->N-1). However, typically, we also provide an initial state.
This results in having two constraints at the initial node. This could be a problem when both of them are active.

Fix:
- enforce path constraints at nodes (1->N-1)
- create a new variable (con_h_expr_0) to set a constraint expression at node 0 explicitly (in case it is desired)
- added a minimal example to reproduce the issue and show usage of the newly added                                                  
  variable (ocp_example_h_init_contraints.py)

### CODE REVIEW

File: interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c [hunk 1]
```diff
@@ -238,6 +239,7 @@ ocp_nlp_dims* {{ model.name }}_acados_create_2_create_and_set_dimensions({{ mode
     int* nsphi = intNp1mem + (N+1)*8;
     int* ns    = intNp1mem + (N+1)*9;
     int* ng    = intNp1mem + (N+1)*10;
+    // int* nh    = intNp1mem + (N)*11;
```
Comment (FreyJo) [inline interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c:242]: remove

**Review by FreyJo — COMMENTED** [review]
Hi Mohammed,
It makes a lot of sense to separate these as discussed!
Thanks for work! :+1: 
I added two comments.
But I think the biggest work remaining is to make the Matlab/Octave tests pass.

File: interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c [hunk 2]
```diff
@@ -2264,6 +2315,14 @@ int {{ model.name }}_acados_update_params({{ model.name }}_solver_capsule* capsu
     {%- endif %}
     {%- endif %}
 
+    {% if constraints.constr_type_0 == "BGH" and dims.nh_0 > 0 %}
+        capsule->nl_constr_h_0_fun_jac.set_param(&capsule->nl_constr_h_0_fun_jac, p);
+        capsule->nl_constr_h_0_fun.set_param(&capsule->nl_constr_h_0_fun, p);
+    {% endif %}
+    {%- if solver_options.hessian_approx == "EXACT" %}
```
Comment (FreyJo) [inline interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c:2322]: this if has to go inside the one before

File: interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c [hunk 3]
```diff
@@ -2399,6 +2458,14 @@ int {{ model.name }}_acados_update_params_sparse({{ model.name }}_solver_capsule
     {%- endif %}
     {%- endif %}
 
+    {% if constraints.constr_type_0 == "BGH" and dims.nh_0 > 0 %}
+        capsule->nl_constr_h_0_fun_jac.set_param_sparse(&capsule->nl_constr_h_0_fun_jac, n_update, idx, p);
+        capsule->nl_constr_h_0_fun.set_param_sparse(&capsule->nl_constr_h_0_fun, n_update, idx, p);
+    {% endif %}
+    {%- if solver_options.hessian_approx == "EXACT" %}
```
Comment (FreyJo) [inline interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c:2465]: this if has to go inside the one before

**Comment by mohammed-hababeh** [comment]
oh, you are absolutely right. Thanks for spotting it. I also found other mistakes, I fixed them. I ran a couple of tests manually, and they passed. Let us see how it goes with the rest.

**Comment by sandmaennchen** [comment]
Hi Mohammed, we recently added the dimension `ns_0` (total number of slacks at stage 0). Might be relevant for your pull request.
https://github.com/acados/acados/pull/963/commits/a11ff48db427cd58b3007e29156e533d3af71ebb

**Comment by FreyJo** [comment]
Hey @mohammed-hababeh
I decided to finish up the work in this PR in https://github.com/acados/acados/pull/994 before there are more conflicts.
I squashed your commits and rebased.
Hope there is not too much left to do :)

**Pull request closed without merging**